Add container start retry

MMaiero opened the pull request:
Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

**Related Issue:** This PR fixes/closes {issue number}

**Description of the solution adopted:** A more detailed description of the changes made to solve/close one or more issues. If the PR is simple and easy to inderstand this section can be skipped.

**Screenshots:** If applicable, add screenshots to help explain your solution

**Any side note on the changes made:** Description of any other change that has been made, which is not directly linked to the issue resolution [e.g. Code clean up/Sonar issue resolution]

MMaiero committed:
Added container start retry policy

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Started refactoring and callback

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero commented:
@nicolatimeus @GregoryIvo Initial work on retries and API review.
APIs will need better javadocs. Tests completely need adaptation
Please let me know if the approach works for you and if you already see some needed improvements

GregoryIvo commented:
@MMaiero I like the direction you are going.  From what I am gathering, you are trying to move out all the 'registration logic' to its own separate layer? API will only be concerned about performing actions on containers, and the DockerServiceListener handles life-cycle tracking?

GregoryIvo commented on kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/DockerService.java:
```diff
@@ -22,34 +22,44 @@
      * @return
      */
-    public String[] listContainersByID();
+    public List<String> listContainersIds();
```
There are many Legacy API functions like this that were created before ContainerDescriptor was realized. I think that going forward we can remove all of these, and rely only on the ContainerDescror API methods.

MMaiero replied:
Don't know. We were discussing with Nicola and maybe it makes sense to have those generic commands that allow to list and run also containers that are not managed by the framework.

GregoryIvo commented on kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/DockerService.java:
```diff
@@ -22,34 +22,44 @@
+    * @param timeOutSecconds
+    * @throws KuraException
+    */
+   public void pullImage(String imageName, String imageTag, int timeOutSecconds) throws KuraException;
```
Going with the theme of my previous comment above, perhaps we add all of the timeOut Parameters into the ContainerDescriptor?

MMaiero replied:
Agreed. Can you do that?

MMaiero committed:
Updated container start and stop logic.

Started updating tests

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero commented:
@GregoryIvo Can you help me with the tests review?
In particular, can you please update the tests in DockerServiceImplTest? I don't want to mess with the logic you used there. Thanks.

MMaiero committed:
More tests updates

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
More APIs changes.

Splitted container stop and delete

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus commented on kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/ContainerDescriptor.java:
```diff
@@ -37,9 +37,9 @@
     private Map<String, String> containerVolumes;
-    private Boolean containerPrivilaged;
+    private Boolean containerPrivileged;
     private ContainerStates containerState = ContainerStates.STOPPING;
```
Keeping the containerState field mutable can be error prone, it requires a lot of internal housekeeping and it can become inconsistent with system state.
What if we just remove it from here and add a method in DockerService that requests the container state to the container daemon given a container id?
Making ContainerDescriptor immutable would also be a nice imporvement.

nicolatimeus commented on kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/impl/DockerServiceImpl.java:
```diff
@@ -120,93 +118,70 @@ public void updated(Map<String, Object> properties) {
+        return container.getImage().split(":")[0];
     }
 
     private int[] parseExternalPortsFromDockerPs(ContainerPort[] ports) {
```
This method will always return an empty array, the same applies to parseInternalPortsFromDockerPs, ArrayUtils.add does not modify the passed array argument, which is immutable.

nicolatimeus replied:
The `ipFormatTest.equals("::")` check fails on my device, the reported id in that case is `0.0.0.0`.
Should we expose the IP in our APIs?

MMaiero committed:
Moved apis to apis bundle.

Other improvements

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixed dockerservice references

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Removed direct container state management

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixed wrong containerId reference

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
ContainerDescriptor now is immutable

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus commented on kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/impl/DockerServiceImpl.java:
```diff
@@ -271,97 +245,141 @@ public String getContainerIDbyName(String name) {
     @Override
-    public void pullImage(String imageName) {
-        pullImage(imageName, 120);
+    public void stopContainer(ContainerDescriptor container) {
```
What about just removing this method? The only field that it consider in the `ContainerDescriptor` is the container name (previously the container id). 
Probably the `stopContainer(String id)` is the minimum required functionality here, the clients can figure out the id given a name/container descriptor using the other methods (this is what the current implementation does). The same applies to `deleteContainer(ContainerDescriptor containerDescriptor)`.

MMaiero committed:
Renamed ConfigurableGenericDockerService

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Removed unneeded methods with ContainerDescriptor

Added javadoc

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Javadoc cleanup

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Image download timeout configurable

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero commented:
Still not sure if we should change DockerService name to something different. Maybe ContainerOrchestrationService is better? Even though the library we use calls everything Docker...

MMaiero committed:
Updated container instance container lifecycle

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

GregoryIvo commented:
> 

I agree with the renaming.  In the future, we can add support for other tools that are similar to docker like podman

nicolatimeus committed:
Added ability to interrupt image pull
Changed ContainerInstance lifecycle
Removed container name from DockerService.registerListener

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Fixes in DockerServiceImpl

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Restored disabled tests

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Updated port parsing check

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Split ContainerDescriptor wip

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

GregoryIvo committed:
fixed and added more tests

MMaiero committed:
Minor renamings in api

Renewed equals and hash methods in API pojos

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Renamed dockerservice to ContainerOrchestrationService

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixed reference in tests

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
More renamings

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
Improved ContainerInstance behavior in startup with already existing container

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

MMaiero committed:
Small name change

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
Moved existing container check to container orchestrator

Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>
